### ISSUE TITLE
Picture problems of start, end and grass

### ISSUE THREAD

**Issue by Shame-fight** [Sana-078/Pathfinding-Visualizer#2]
Thanks for your work, I want to replace the corresponding squares of start, end and grass on the map with pictures instead of filling like red, purple or green. The goal is that the effect will look more realistic. But I'm new to pygame, trying to modify the code but can't achieve what I want. can you help me? Is it possible to provide an example such as modifying the picture of start.
Looking forward to hearing from you.
@Sana-078

**Comment by Sana-130** [comment]
As u asked, I modified the code and this is how it looks ,
![pfpic](https://user-images.githubusercontent.com/70009983/156172415-6bd70a88-4362-459d-9f5f-b52b0f81831e.png)

Also since we are representing the green area with a grass image I thought it would be a good idea to give them a fixed weight rather than a random value since the user wont be able to see it. Right now , the start and end points contain a flag image , u can change it to whatever u want, just the image would be scaled to the size of that cell, before displaying. Let me know if there is anything else u want to add @Shame-fight .

**Comment by Shame-fight** [comment]
Thanks for your reply, I want to do the following things.
1. Add a variety of obstacles. At present, the program supports one obstacle: grass, whether it is possible to add a variety of different obstacles, such as hills, lakes, swamps, etc.
2. Support three algorithms, including A*, Dijkstra and BFS algorithms. One is currently supported.
3. Multiple start and end pairs. Simulate multiple cars running the path planning algorithm at the same time.
4. Dynamic display of paths. There is currently an orange path between the start and end points. Whether the path changes can be displayed dynamically, for example, there is a car icon that goes from the starting point to the ending point step by step, so that the final path can be displayed more conveniently.
5. Save and load maps. Whether the constructed map can be saved, including obstacles, etc. If there is this function, you can load the saved map, select the starting point, the ending point, add obstacles, etc. without building a new map every time.
3rd and 4th if it's more difficult to implement, can you implement 1st and 2nd?
Thanks again for your work.

@Sana-078

**Comment by Shame-fight** [comment]
The first requirement I have done, by binding keyboard events。
I want to finish the second and third points, can you help me。
@Sana-078

**Comment by Sana-078** [comment]
Sorry i took some time to reply, was caught up in another project. Anyway I tried to implement the dynamic display of paths.. (it's a really cool idea btw) and it's not really working the way it's supposed to.. anyway I will checkup on the others and will let u know.

**Comment by Sana-130** [comment]
Sorry i had a hard time replying, was caught up in another project. Anyway I tried to implement the dynamic display of paths.. (it's a really cool idea btw) and it's not really working the way it's supposed to..  about the rest, I will check it out and will let u know  @Shame-fight .

**Comment by Sana-078** [comment]
Dynamic display of paths, finally worked, u can check that out.
 
![Screenshot (139)](https://user-images.githubusercontent.com/70009983/159449962-a6eb17ba-bf21-4e8c-9a58-0feadd3a9c79.png)

About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading??

**Comment by Sana-130** [comment]
Dynamic display of paths, finally worked, u can check that out.

About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight

**Comment by Shame-fight** [comment]
> Dynamic display of paths, finally worked, u can check that out.
> 
> About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight

Thanks for your time, the dynamic effect is cool.
Have you ever thought about achieving the following effect:
Instead of generating a dynamic path at the end, the cart moves to the corresponding grid with each step. This path may not be optimal, but it can truly reflect the path planning of unmanned vehicles in reality.
@Sana-078

**Comment by Shame-fight** [comment]
Can you add the A*s algorithm, I tried to add this algorithm, but the final path does not consider the high pass cost of the grass, the algorithm treats the grass as a general path, I don't know how to solve this problem.
@Sana-078

**Comment by Sana-078** [comment]
> Can you add the A*s algorithm, I tried to add this algorithm, but the final path does not consider the high pass cost of the grass, the algorithm treats the grass as a general path, I don't know how to solve this problem. @Sana-078

yeah I have added it, check the program. @Shame-fight

**Comment by Sana-130** [comment]
> Can you add the A*s algorithm, I tried to add this algorithm, but the final path does not consider the high pass cost of the grass, the algorithm treats the grass as a general path, I don't know how to solve this problem. @Sana-078

Yeah I have added it. Check the code. @Shame-fight

**Comment by Sana-130** [comment]
> > Dynamic display of paths, finally worked, u can check that out.
> > About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight
> 
> Thanks for your time, the dynamic effect is cool. Have you ever thought about achieving the following effect: Instead of generating a dynamic path at the end, the cart moves to the corresponding grid with each step. This path may not be optimal, but it can truly reflect the path planning of unmanned vehicles in reality. @Sana-078

Care to elaborate the last part.

**Comment by Shame-fight** [comment]
> > > Dynamic display of paths, finally worked, u can check that out.
> > > About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight
> > 
> > 
> > Thanks for your time, the dynamic effect is cool. Have you ever thought about achieving the following effect: Instead of generating a dynamic path at the end, the cart moves to the corresponding grid with each step. This path may not be optimal, but it can truly reflect the path planning of unmanned vehicles in reality. @Sana-078
> 
> Care to elaborate the last part.

In the real world, the path planning of the unmanned vehicle should be like this: the unmanned vehicle only walks one grid, and then decides the grid position to go next according to the current algorithm and the position of the surrounding obstacles. Repeat this cycle until you reach the end.
At present, our implementation method is: after finding the path of the starting point and the ending point, select the shortest path.
So, is there a possibility to take the shortest path currently implemented as the "reference answer", the actual path is that the unmanned vehicle only walks one grid at a time, and iterates through this, and finally reaches the end point, of course, this path is very large It may not be a "reference answer", and the actual path obtained from each run should be different, because we have four grids to go in each step.
The final effect: two paths are displayed, the first is the shortest path, that is, the "reference answer"; the other is the actual path displayed step by step.
These are my thoughts, thanks for your work!
@Sana-078

**Comment by Shame-fight** [comment]
> Dynamic display of paths, finally worked, u can check that out.
> 
> About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight

Yes, there are multiple autonomous vehicles on the same map. Different cars can choose different starting and ending points and algorithms, and finally start path planning at the same time.

**Comment by Sana-130** [comment]
> > Dynamic display of paths, finally worked, u can check that out.
> > About the algorithm, I'm not really sure how bfs would work on a weighted graph. And by the third, did u mean threading?? @Shame-fight
> 
> Yes, there are multiple autonomous vehicles on the same map. Different cars can choose different starting and ending points and algorithms, and finally start path planning at the same time.

It's possible but the issue is that the GUI is going to glitch a lot even though it's just one thread that's currently running. @Shame-fight
